airbyte-cdk/python/airbyte_cdk/sources/availability_strategy.py

```diff
@@ -0,0 +1,145 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+import requests
+from abc import ABC, abstractmethod
+from requests import HTTPError
+from typing import Dict, List, Optional, Text, Tuple
+
+from airbyte_cdk.models.airbyte_protocol import SyncMode
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources import Source
+
+
+class AvailabilityStrategy(ABC):
+    """
+    Abstract base class for checking stream availability
+    """
+
+    @abstractmethod
+    def check_availability(self, source: Source, logger: logging.Logger, stream: Stream) -> Tuple[bool, any]:
+        """
+        Checks stream availability.
+
+        :param source: source
+        :param logger: source logger
+        :param stream: stream
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+
+
+class HTTPAvailabilityStrategy(AvailabilityStrategy):
+    def check_availability(self, source: Source, logger: logging.Logger, stream: Stream) -> Tuple[bool, str]:
+        """
+        Check stream availability by attempting to read the first record of the
+        stream.
+
+        :param source: source
+        :param logger: source logger
+        :param stream: stream
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+        try:
+            # Some streams need a stream slice to read records (e.g. if they have a SubstreamSlicer)
+            stream_slice = self._get_stream_slice(stream)
+            records = stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=stream_slice)
+            next(records)
+        except HTTPError as error:
+            return self.handle_http_error(source, logger, stream, error)
+        return True, None
+
+    def handle_http_error(self, source: Source, logger: logging.Logger, stream: Stream, error: HTTPError):
+        """
+        Override this method to define error handling for various `HTTPError`s
+        that are raised while attempting to check a stream's availability.
+
+        :param source: source
+        :param logger: source logger
+        :param stream: stream
+        :param error: HTTPerror raised while checking stream's availability.
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+        if error.response.status_code == requests.codes.FORBIDDEN:
+            error_message = "This is most likely due to insufficient permissions on the credentials in use. "
```

Comment:
What do you think about mentioning we considered unavailability because we received a `403` HTTP response?